### PR TITLE
Pootle fs boilerplate (/more)

### PR DESCRIPTION
**Pull request by phlax**
Adds

- a cleanup/fix for update/sync stores
- pootle_fs decorators
- pootle_fs util for getting plugin
- a fix and some additional tests for pootle_fs.FSFile

### CODE REVIEW

File: tests/pootle_fs/utils.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    def plugin_method(self, foo):
+        return "Plugin method called with: %s" % foo
+
+
+@pytest.mark.django_db
+def test_project_fs_instance():
+
+    @provider(fs_plugins, sender=Project)
+    def provide_fs_plugin(**kwargs):
```
Comment (unho) [inline tests/pootle_fs/utils.py]: Defined and unused?

File: tests/pootle_fs/utils.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    project.config["pootle_fs.fs_type"] = None
+    project.config["pootle_fs.fs_url"] = "bar"
+    with pytest.raises(NotConfiguredError):
+        FSPlugin(project)
+    project.config["pootle_fs.fs_type"] = "foo"
+    with pytest.raises(MissingPluginError):
+        FSPlugin(project)
+
+    @provider(fs_plugins, sender=Project)
+    def provide_fs_plugin(**kwargs):
```
Comment (unho) [inline tests/pootle_fs/utils.py]: Defined and unused too?

**Comment by unho** [comment]
Other than the doubts, lgtm.

File: tests/pootle_fs/utils.py (hunk 1 above)
Comment (phlax) [inline tests/pootle_fs/utils.py]: on the contrary

File: tests/pootle_fs/utils.py (hunk 2 above)
Comment (phlax) [inline tests/pootle_fs/utils.py]: @unho these are plugin hookups that are necessary to make the test work

File: pootle/apps/pootle_fs/files.py [hunk 3]
```diff
@@ -205,13 +205,18 @@ def _sync_from_pootle(self):
             f.write(self.serialize())
         logger.debug("Pushed file: %s" % self.path)
 
-    def _sync_to_pootle(self, merge=False, user=None):
+    def _sync_to_pootle(self, merge=False, user=None, pootle_wins=None):
```
Comment (ta2-1) [inline pootle/apps/pootle_fs/files.py]: `None` vs `False`?

Comment (phlax) [inline pootle/apps/pootle_fs/files.py]: thats correct - True/False have a different meaning

File: tests/models/store_fs.py [hunk 4]
```diff
@@ -228,3 +230,47 @@ def test_save_store_fs_bad_project(tp0_store_fs):
...
+        file_class = FSFile
+
+        def __init__(self, project):
+            self.project = project
+
+        def foo(self):
+            return "bar"
+
+    @provider(fs_plugins)
+    def provide_plugin(**kwargs):
```
Comment (ta2-1) [inline tests/models/store_fs.py]: how is it used?

Comment (phlax) [inline tests/models/store_fs.py]: this is a dummy version of the FSPlugin.

the provider hooks it up, and lets us test for the file and plugin @property on StoreFS

File: pootle/apps/pootle_fs/files.py (hunk 3 above)
Comment (ta2-1) [inline pootle/apps/pootle_fs/files.py]: k, but it'd be helpful to see some real cases. I assume you had some reasons when added `pootle_wins` param.

File: tests/models/store_fs.py (hunk 4 above)
Comment (ta2-1) [inline tests/models/store_fs.py]: cool 👍

File: pootle/apps/pootle_fs/files.py (hunk 3 above)
Comment (phlax) [inline pootle/apps/pootle_fs/files.py]: yes.

It works like this
- sometimes there *is* a store_fs and a previous action was required (conflict)
- there is a store_fs but no previous action required - and then pull is called with the pootle_wins param accordingly
- there is no store_fs - pull is called with the appropriate pootle_wins param

**Comment by ta2-1** [comment]
@phlax thanks. all looks good.
lgtm

**Pull request merged**